File: packages/models/src/models/resource_reference.py
import re
from dataclasses import dataclass
from typing import Literal, cast, get_args

from models.errors import InvalidError

ResourceType = Literal[
    "device", "driver", "transport", "command", "automation", "fault", "asset"
]

_RESOURCE_TYPES: frozenset[str] = frozenset(get_args(ResourceType))

_URI_PATTERN = re.compile(r"^resource://([^/]+)/(.+)$")
"""Matches resource://<resource_type>/<resource_id> — e.g. resource://device/abc123"""


@dataclass(frozen=True)
class ResourceReference:
    resource_type: ResourceType
    resource_id: str

    def serialize(self) -> str:
        return f"resource://{self.resource_type}/{self.resource_id}"

    @classmethod
    def parse(cls, s: str) -> "ResourceReference":
        match = _URI_PATTERN.match(s)
        if not match:
            msg = f"Invalid resource reference URI: {s!r}"
            raise InvalidError(msg)
        resource_type, resource_id = match.group(1), match.group(2)
        if resource_type not in _RESOURCE_TYPES:
            msg = f"Unknown resource type: {resource_type!r}"
            raise InvalidError(msg)
        return cls(
            resource_type=cast("ResourceType", resource_type),
            resource_id=resource_id,
        )

File: apps/ui/src/lib/resourceReference.ts
export const RESOURCE_TYPES = [
  "device",
  "driver",
  "transport",
  "command",
  "automation",
  "fault",
  "asset",
] as const;

export type ResourceType = (typeof RESOURCE_TYPES)[number];

export type ResourceReference = {
  type: ResourceType;
  id: string;
};

const URI_PATTERN = /^resource:\/\/([^/]+)\/(.+)$/;

export function parseResourceReference(uri: string): ResourceReference | null {
  const match = URI_PATTERN.exec(uri);
  if (!match) return null;
  const [, rawType, id] = match;
  if (!(RESOURCE_TYPES as readonly string[]).includes(rawType)) return null;
  return { type: rawType as ResourceType, id };
}

export function serializeResourceReference(ref: ResourceReference): string {
  return `resource://${ref.type}/${ref.id}`;
}

export function resourceTypeToPath(type: ResourceType, id: string): string {
  switch (type) {
    case "device":
      return `/devices/${id}`;
    case "driver":
      return `/drivers/${id}`;
    case "transport":
      return `/transports/${id}`;
    case "asset":
      return `/assets/${id}`;
    case "automation":
      return `/automations/${id}`;
    case "fault":
      return "/faults";
    case "command":
      return "/devices/commands";
  }
}
